add : undo/redo system for  editor

Sahelisaha04 opened the pull request:
https://github.com/user-attachments/assets/8658a295-c302-44f2-b711-b2ac473ece6f


Fixes #707
/claim #707

Sahelisaha04 committed:
Add cursor-aware undo/redo system for trigger editor

- Implement priority-based undo/redo based on cursor position
- Lua editor focused: only undos Lua editor changes (edbee)
- Text fields focused: only undos text field changes (unified TextChangeCommand)
- Tree/UI focused: only undos item operations (global stack)
- Add cursor position preservation during text undo/redo
- Support for 79 QLineEdit and 51 QPlainTextEdit widgets
- Up to 50 undo levels with proper command merging
- Both keyboard shortcuts (Ctrl+Z/Ctrl+Y) and toolbar buttons
- Complete Google Docs-like sequential undo/redo experience

Fixes #707

Sahelisaha04 commented:
@vadi2 is my approach correct ?

Sahelisaha04 committed:
Fix CodeFactor style issues in TriggerEditorCommands

- Fix else clause formatting to appear on same line as preceding }
- Add consistent brace formatting for all else clauses
- Add missing newlines at end of both .cpp and .h files
- Resolves all 15 CodeFactor style issues

All else statements now follow proper C++ style guidelines:
- else appears on same line as preceding }
- Consistent brace placement throughout
- Proper file ending with newlines

Sahelisaha04 committed:
Merge branch 'development' into feature/707-cursor-aware-undo-redo

vadi2 commented:
Yep you definitely are on the right path. Just make sure that everything that you see as toggleable/settable in the editor (for example, the right side of the trigger pattern) can be undone as well. Also avoid using while loops if possible, change them to for loops if logic allows it.

Sahelisaha04 committed:
Fix build error: Add missing Qt widget includes

Added missing #include directives for QCheckBox, QSpinBox, and QComboBox
to TriggerEditorCommands.h to resolve compilation errors.

Sahelisaha04 committed:
Revert "Fix build error: Add missing Qt widget includes"

This reverts commit 4da28e1d40fc8e1594445513a913aaddff2fbe62.

Sahelisaha04 committed:
Implement comprehensive undo/redo system for GitHub issue #707

- Add CheckboxChangeCommand, SpinboxChangeCommand, ComboboxChangeCommand for UI element undo support
- Implement cursor-aware undo/redo system that switches between text editor and UI element undo based on focus
- Add setupUndoForUIElements() method to systematically connect all toggleable UI elements
- Replace iterator-based while loops with modern for loops for better code clarity
- Add 50-level undo stack with command merging capabilities
- Support undo for all major UI controls: checkboxes, spinboxes, comboboxes in trigger editor
- Enhance undo button tooltips to show context-specific information
- Add proper signal blocking to prevent recursion during undo/redo operations

Sahelisaha04 commented:
https://github.com/user-attachments/assets/806ff80f-f781-4f69-8434-ac6a304c1ef2


https://github.com/user-attachments/assets/8767015f-53a6-48e0-ba96-55b222c1410f


cc @vadi2  i have made the changes

ZookaOnGit commented:
couple of issues found;

1. open Mudlet / Script editor, click on an element (trigger, alias, whatever), click undo.. stuff starts disappearing but I haven't changed anything yet
2. create a script, delete it, click undo, ALL scripts are duplicated

Sahelisaha04 committed:
Implement intelligent unified undo/redo system for cross-element functionality

This commit resolves the issue where undo/redo was isolated per element type,
preventing users from undoing changes made in triggers while viewing scripts.

Key improvements:
- Cross-element undo capability: Can now undo trigger changes while in scripts tab
- Intelligent priority system: Fast Lua editor undo + unified stack for major changes
- Fixed spurious undo commands on item selection
- Fixed script deletion undo duplication issues
- Smart button states showing combined availability from all stacks

Technical changes:
- Added LuaEditorChangeCommand for unified Lua editor integration
- Enhanced all slot_*Selected functions with proper state management
- Improved populateScripts() to prevent undo tracking during population
- Made mCreatingFromUndoCommand and mIsLoadingItem public for command access
- Added intelligent priority logic that maintains context while enabling cross-element undo

Fixes GitHub issue #707: Users can now seamlessly undo/redo across all element types
while maintaining optimal performance for rapid text editing scenarios.

Sahelisaha04 commented:
@ZookaOnGit now its okay ?


https://github.com/user-attachments/assets/0ff6f733-7104-4de0-85bc-74ad939aa5a1


https://github.com/user-attachments/assets/f8210b46-9fad-40cc-bd4b-c7e30d2f2744

Sahelisaha04 committed:
Remove TODO and simplify LuaEditorChangeCommand

Remove TODO comment about caret positioning and simplify the implementation
to focus on text content restoration only. Caret positioning for edbee
editor is complex and not essential for the core cross-element undo
functionality.

Changes:
- Removed caret position tracking variables (mOldCaretPos, mNewCaretPos)
- Simplified restoreState() to only handle text content
- Updated function signatures to remove unused caretPos parameter
- Cleaned up captureCurrentState() method

Sahelisaha04 commented:
@vadi2 @ZookaOnGit just friendly ping , can you pls tell me why the windows lua test failed ?  how to fix it ?

Sahelisaha04 committed:
Merge branch 'development' into feature/707-cursor-aware-undo-redo

vadi2 commented:
Compiling Mudlet failed, that's why the tests couldn't run. I'm on mobile right now, but have a look at the end of the compile job.

ZookaOnGit commented:
you can check the build logs by clicking on the offending failure; seems to be failing here

https://github.com/Mudlet/Mudlet/actions/runs/17518307697/job/49758727136?pr=8167#step:6:712

edit: those github generated links don't work apparently - check the Windows Build job, right at the end

Sahelisaha04 committed:
Fix Windows build: Add TriggerEditorCommands to mudlet.pro

The Windows CI build was failing with undefined reference errors for
AddItemCommand, DeleteItemCommand, and DeleteMultipleItemsCommand classes.
These classes are defined in TriggerEditorCommands.cpp/h which were added
to CMakeLists.txt but not to mudlet.pro (used for QMake builds on Windows).

This commit adds both TriggerEditorCommands.cpp and TriggerEditorCommands.h
to the appropriate SOURCES and HEADERS sections in mudlet.pro.

Sahelisaha04 committed:
Merge branch 'development' into feature/707-cursor-aware-undo-redo

Sahelisaha04 commented:
@vadi2 @ZookaOnGit all the test passed could you pls test it

ZookaOnGit committed:
Merge branch 'development' into feature/707-cursor-aware-undo-redo

ZookaOnGit commented:
Can't undo;
- timers Time values (hours, mins, secs, ms)
- trigger highlighting colour changes/checkbox
- trigger types (substring, perl regex, etc)
- button rotation field

then this happened; the previous timer somehow ended up in my trigger section

<img width="2735" height="1733" alt="Screenshot_20250909_103734-1" src="https://github.com/user-attachments/assets/e448ab61-ab70-4092-826d-67b53889e546" />

seems to happen with timer groups
- create new timer group
- move to triggers, undo - timer group appears in triggers secton

Sahelisaha04 commented:
https://github.com/user-attachments/assets/f822e595-a19a-40d1-9c0e-7e1249c8ea8a

@ZookaOnGit  fixed

Sahelisaha04 commented:
@vadi2 @ZookaOnGit just a friendly ping

ZookaOnGit commented:
- trigger highlighting colour changes/checkbox not done,
- trigger groups don't undo/redo,
- undo/redo gets confused if I cancel from performing an undo which would delete a group.  A redo makes a new group.

https://github.com/user-attachments/assets/c77becb2-cfad-4adf-aca3-30a14cc6fb44

Please test thoroughly before requesting another review.

Sahelisaha04 committed:
Merge development branch and complete undo/redo system for trigger editor

Sahelisaha04 committed:
Merge branch 'development' into feature/707-cursor-aware-undo-redo

Sahelisaha04 committed:
Fix trigger visibility after undo restoration

Fixed issue where deleted triggers were restored in memory but not visible
in the UI tree widget after undo operations. The problem was that
populateTriggers() was commented out for triggers while other item types
(timers, aliases, etc.) properly refreshed their tree widgets.

Uncommented populateTriggers() call in TriggerEditorCommands.cpp line 202
to ensure triggers become visible after undo restoration, matching the
behavior of all other item types.

Sahelisaha04 commented:
https://github.com/user-attachments/assets/a0774e7f-e5db-4247-911d-94b27fda734c


https://github.com/user-attachments/assets/9203e6ca-3d5f-46c9-9152-8acc0d4161fa



@ZookaOnGit @vadi2 i am almost ther few more changes left

Sahelisaha04 committed:
Merge branch 'development' into feature/707-cursor-aware-undo-redo

vadi2 commented:
Solved in another bounty.
